### PR TITLE
[WIP]Implement DeviceContextManager and Ensure only one CUDA stream existing in new framework at now

### LINKED ISSUE #3796: how to manage DeviceContext which contains cuda streams
When launching a kernel in CUDA, we have to set a specific CUDA stream. We have implement DeviceContext which has a CUDA stream. Copy and operator run job will launch a CUDA kernel.   So we have to pass a DeviceContext parameter to Copy and Operator job.
In user end, users will only set CPU or several GPU ids to run network training, they are not aware of  DeviceContext. So, we have to implement a DeviceContextManager to initialize some DeviceContext at first, and schedule the CDUA kernels.

And OperatorBase may need another run interface,
```
class OperatorBase {
  void Run(Place p);
  void Run(Scope, DeviceContext);
};
``` 
The first Run interface is to users, they will only need to set a place, and paddle  will pass the DeviceContext to the second Run interface.

It's the  same with copy jobs. A specific CUDA stream will pass to a copy job and it's scheduled by DeviceContextManager.

### PR DESCRIPTION
**Pull request by QiJune**
Fix #3796

### CODE REVIEW

File: paddle/framework/tensor.h [hunk 1]
```diff
@@ -94,13 +94,15 @@ class Tensor {
...
+   * @param[in] src         The external tensor.
+   * @param[in] dst_place   Destination Place copy to.
+   * @param[in] is_sync     Flag for sync or async
    *
    * @note    CopyFrom supports CPU <-> GPU, GPU <-> GPU.
    */
   template <typename T>
-  inline void CopyFrom(const Tensor& src, const platform::Place& dst_place);
+  inline void CopyFrom(const Tensor& src, const platform::Place& dst_place,
+                       bool is_sync = false);
```
Comment (reyoung) [inline paddle/framework/tensor.h]: I am not sure, but the easiest way might be to take a `const DeviceContext&` by `CopyFrom`?

**Review by reyoung — CHANGES_REQUESTED** [review]
I do not think `DeviceContextMgr` is necessary. Maybe we could just pass a `DeviceContext` to `Tensor::CopyFrom` and make it async?

File: paddle/framework/tensor.h (hunk 1 above)
Comment (QiJune) [reply paddle/framework/tensor.h]: It's hard for users to pass a DeviceContext to CopyFrom method. Users have to pass right DeviceContext depends on the src place(cpu or gpu) and dst place(cpu or gpu).

**Comment by QiJune** [comment]
@reyoung
 
1.  DeviceContext should not expose to users, and users should not and almost could not create right DeviceContext. If we pass a DeviceContext to CopyFrom, so where is the DeviceContext created? 
2. We also use copy method in pybind between C++ and python, and we can hardly pass a DeviceContext here, and expose a api like this:
```
a = numpy.array(1, 2)
t.set_float(a, device_context);
```
3. Anyway, we need to create some DeviceContext in Paddle initialization stage, and handle the parallel CUDA  streams internally.

**Pull request closed without merging**